### PR TITLE
imposeOrderFlowFee Tests

### PR DESCRIPTION
**Pull request by Flocqst**
Add `OrderFlowFee` logic for Delayed orders, add Event on `OrderFlowFeeImposed`, add tests for both `setOrderFlowFee` and `_imposeOrderFlowFee` functionnalities.

## Description
* Overloads `_imposeOrderFlowFee` and `_calculateOrderFlowFee` with `uint256 _desiredFillPrice`, which is set only for delayedOrders (setting a Delayed order on desired price 0 would result in paying the fee at the current price, but this edge case is not relevant as having 0 as fill price is irrelevant)
* Add `emitOrderFlowFeeImposed` event logic
* Add tests for `test_imposeOrderFlowFee`

## Motivation and Context
Part of KIP-122

### CODE REVIEW

File: src/Account.sol [hunk 1]
```diff
@@ -635,32 +645,56 @@ contract Account is IAccount, Auth, OpsReady {
...
+        uint256 _desiredFillPrice
+    ) internal view returns (uint256) {
         // fetch order flow fee from settings
         uint256 orderFlowFee = SETTINGS.orderFlowFee();
 
         // fetch current sUSD exchange rate for market
         IPerpsV2MarketConsolidated market = IPerpsV2MarketConsolidated(_market);
         (uint256 price,) = _sUSDRate(market);
 
+        uint256 usedPrice = (_desiredFillPrice != 0) ? _desiredFillPrice : price;
```
Comment (JaredBorders) [inline src/Account.sol:681]: if we use `_desiredFillPrice` there is no need to query `_sUSDRate`, right? lets move this line up and only call `_sUSDRate` if necessary

Comment (JaredBorders) [reply src/Account.sol:681]: something like this
```
uint price;
if (_desiredFillPrice != 0) {
    price = _desiredFillPrice;
} else {
    (price,) = _sUSDRate(IPerpsV2MarketConsolidated(_market));
}
```

File: src/Account.sol [hunk 2]
```diff
@@ -774,8 +808,7 @@ contract Account is IAccount, Auth, OpsReady {
         int256 _sizeDelta,
         uint256 _desiredFillPrice
     ) internal {
-        /// @custom:todo use _desiredFillPrice
-        _imposeOrderFlowFee(_market, _sizeDelta);
+        _imposeOrderFlowFee(_market, _sizeDelta, _desiredFillPrice);
```
Comment (JaredBorders) [inline src/Account.sol:811]: shouldn't we use `_imposeOrderFlowFee(_market, _sizeDelta, _desiredFillPrice)` for `_perpsV2SubmitDelayedOrder` too since there is a desired fill price specified there too?

File: src/Account.sol
Comment (JaredBorders) [inline src/Account.sol:1]: Should we use desired fill price in the `executeConditionalOrder()` function?

File: test/unit/Settings.t.sol [hunk 3]
```diff
@@ -116,18 +118,26 @@ contract SettingsTest is Test, ConsolidatedEvents {
...
     }
 
     function test_setOrderFlowFee_Event() public {
-        /// @custom:todo
-    }
-
-    function test_setOrderFlowFee_InvalidOrderFlowFee() public {
-        /// @custom:todo
+        vm.expectEmit(true, true, true, true);
+        emit ExecutorFeeSet(5);
```
Comment (JaredBorders) [inline test/unit/Settings.t.sol:140]: In the future, avoid using random values like this. 

I recommend just fuzzing this and `vm.assume(value < MAX_ORDER_FLOW_FEE)`

File: test/unit/Settings.t.sol [hunk 4]
```diff
@@ -116,18 +118,26 @@ contract SettingsTest is Test, ConsolidatedEvents {
     //////////////////////////////////////////////////////////////*/
 
     function test_setOrderFlowFee(uint256 fee) public {
-        /// @custom:todo
+        if (fee > 100_000) {
```
Comment (JaredBorders) [inline test/unit/Settings.t.sol:121]: use `MAX_ORDER_FLOW_FEE` (will need to be pulled from settings)

**Review by JaredBorders — CHANGES_REQUESTED** [review]

File: test/integration/orderFlowFee.behavior.t.sol [hunk 5]
```diff
@@ -107,57 +115,275 @@ contract OrderFlowFeeTest is Test, ConsolidatedEvents {
...
+            assertEq(imposedOrderFlowFee, 0);
+        } else {
+            uint256 orderFlowFeeMath = abs(sizeDelta) * sUSDmarketRate
+                * settings.orderFlowFee() / settings.MAX_ORDER_FLOW_FEE();
+            assertEq(orderFlowFeeMath, imposedOrderFlowFee);
+        }
     }
 
-    /// @custom:todo use atomic order flow to simplify testing
+    /// Verifies that OrderFlowFee is correctly sent from account margin to treasury when there is enough funds in account margin to cover orderFlowFee
```
Comment (JaredBorders) [inline test/integration/orderFlowFee.behavior.t.sol:150]: split this into two lines

File: test/integration/orderFlowFee.behavior.t.sol [hunk 6]
```diff
@@ -107,57 +115,275 @@ contract OrderFlowFeeTest is Test, ConsolidatedEvents {
...
+        // because there is not enough margin for the position delta.
+        vm.expectRevert("Insufficient margin");
 
-    /// @custom:todo use atomic order flow to simplify testing
+        account.execute(commandsAtomic, inputsAtomic);
+
+        uint256 treasuryPostBalance = sUSD.balanceOf(settings.TREASURY());
+
+        // Assert that no overflowfee was distributed
+        assertEq(treasuryPreBalance, treasuryPostBalance);
```
Comment (JaredBorders) [inline test/integration/orderFlowFee.behavior.t.sol:298]: hmm, maybe I am mistaken, but if a revert is caught and asserted, will the following code be reached??

File: test/integration/orderFlowFee.behavior.t.sol [hunk 7]
```diff
@@ -107,57 +115,275 @@ contract OrderFlowFeeTest is Test, ConsolidatedEvents {
...
+        (uint256 desiredFillPrice,) =
+            IPerpsV2MarketConsolidated(market).assetPrice();
+
+        vm.expectEmit(true, true, true, true);
+        // orderFlowFee is 94025250000000000 in this configuration
+        emit OrderFlowFeeImposed(address(account), 94_025_250_000_000_000);
+
+        submitAtomicOrder(sETHPERP, marginDelta, sizeDelta, desiredFillPrice);
+    }
+
```
Comment (JaredBorders) [inline test/integration/orderFlowFee.behavior.t.sol:320]: would be good to have these tests where there is a pending conditional order and thus the free margin != actual margin in the account. This is just to make sure "locked" account margin is never used to pay the fee.

**Review by JaredBorders — CHANGES_REQUESTED** [review]
made some comments specific to order flow fee behavior

File: src/Account.sol (hunk 1 above)
Comment (Flocqst) [reply src/Account.sol]: Updated

File: test/integration/orderFlowFee.behavior.t.sol (hunk 7 above)
Comment (Flocqst) [reply test/integration/orderFlowFee.behavior.t.sol:389]: Added test_imposeOrderFlowFee_market_margin_with_pending_order() for this

File: test/unit/Settings.t.sol (hunk 3 above)
Comment (Flocqst) [reply test/unit/Settings.t.sol:140]: Noted

File: src/Account.sol [hunk 8]
```diff
@@ -1058,7 +1058,8 @@ contract Account is IAccount, Auth, OpsReady {
             _amount: conditionalOrder.marginDelta
         });
 
-        _imposeOrderFlowFee(address(market), conditionalOrder.sizeDelta);
+        // Use conditionalOrder.desiredFillPrice as the submitted order is a delayed order
+        _imposeOrderFlowFee(address(market), conditionalOrder.sizeDelta, conditionalOrder.desiredFillPrice);
```
Comment (JaredBorders) [inline src/Account.sol:1062]: just realized this will impose the fee twice

**Review by JaredBorders — CHANGES_REQUESTED** [review]

File: src/Account.sol (hunk 8 above)
Comment (Flocqst) [reply src/Account.sol]: Removed it. Fee is already imposed during `_perpsV2SubmitOffchainDelayedOrder`.

**Review by JaredBorders — APPROVED** [review]
LGTM

Nice work 👌
